### PR TITLE
[GEOT-6889] Configuration of JNDI and exception handling

### PR DESCRIPTION
**Pull request by roarbra**
[GEOT-6889](https://osgeo-org.atlassian.net/browse/GEOT-6889) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

The function getInitialContext threw an IllegalArgumentException that caller didn't catch. The functionality behind new InitialContext is rather mysterious, so I've tried to log those variables that I've seen influence it. That might change by time.

Lastly the argument hints wasn't used, and probably that was the whole problem for me.

# Checklist

Reviewing is a process done by project maintainers, **mostly on a volunteer basis** (thus limited in time). We need to keep the review overhead as small as possible, and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [X] Confirm you have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md)
- [X] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [X] Make sure the first PR targets the `main` branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.
- [X] The changes are not causing two modules to share the same Java packages (to avoid [Java 9+ split package](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed) issues)
- [ ] The changes are not breaking the build in downstream projects using SNAPSHOT dependencies, GeoWebCache and GeoServer (there is an automatic PR check verifying this, check this when it turns green).

The following are required only for core and extension modules (they are welcomed, but not required, for unsupported modules):

- [X] There is an issue in [Jira](https://osgeo-org.atlassian.net/projects/GEOT) describing the bug/task/new feature (a notable exemptions is, changes not visible to end users). The ticket is for the GeoTools project, if the issue was found elsewhere it's a good practice to link to the origin ticket/issue.
- [X] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [ ] PR for bug fixes and small new features are presented as a single commit
- [ ] Commit message(s) must be in the form "[GEOT-XYZW] Title of the Jira ticket"
- [ ] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by Continuous Integration after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html) (QA checks results will be reported by Continuous Integration after opening this PR)
- [ ] Documentation has been updated accordingly.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.

### CODE REVIEW

**Comment by aaime** [comment]
This looks... weird. Hints is designed to contain tons of little things that control the behavior of GeoTools, as one can see from the plenty of [statically defined keys](https://github.com/geotools/geotools/blob/main/modules/library/metadata/src/main/java/org/geotools/util/factory/Hints.java#L72) in the class. Also keys can only be subclasses or ``RenderingHints.Key``, which is another odd thing to have in an InitialContext... actually it seems broken, since ``InitialContext.lookup(Name)`` wants a ``java.naming.Name`` as the lookup key, while ``RenderingHints.Key`` does not even have a name.. not sure how the two can be reconciled.

Hints are not used, but if they were, I would just look into the Hints for special keys controlling the initial context creation, like which implementation to use (subclasses are possible), and maybe have a key for its initial contents.

**Comment by roarbra** [comment]
I've always thought of those hints acting as a String within the Map, but I see now that it isn't.
Since that argument wasn't used, I took the liberty to introduce a function without arguments. That should save someone else from frustration, but I don't know what happened to the original bug.

**Comment by aaime** [comment]
@jodygarnett any opinion on this one?

File: modules/library/metadata/src/main/java/org/geotools/util/factory/GeoTools.java [hunk 1]
```diff
@@ -938,15 +941,109 @@ public static EntityResolver getEntityResolver(Hints hints) {
...
+            if (javaHome != null) {
+                sb.append("Or from a file specified by system property java.home:\n")
+                        .append(javaHome)
+                        .append("\n");
+            }
+            LOGGER.log(Level.WARNING, sb.toString());
+        }
+        if (e instanceof NamingException) {
+            return (NamingException) e;
+        } else {
```
Comment (groldan) [inline modules/library/metadata/src/main/java/org/geotools/util/factory/GeoTools.java:1040]: Unnecessary else clause.

File: modules/library/jdbc/src/main/java/org/geotools/data/jdbc/datasource/JNDIDataSourceFactory.java [hunk 2]
```diff
@@ -64,8 +64,7 @@ public boolean canProcess(Map<String, ?> params) {
     public DataSource createNewDataSource(Map<String, ?> params) throws IOException {
         String refName = (String) JNDI_REFNAME.lookUp(params);
         try {
-            return (DataSource)
-                    GeoTools.getInitialContext(GeoTools.getDefaultHints()).lookup(refName);
+            return (DataSource) GeoTools.getInitialContext().lookup(refName);
         } catch (Exception e) {
             throw new DataSourceException("Could not find the specified data source in JNDI", e);
         }
```
Comment (groldan) [inline modules/library/jdbc/src/main/java/org/geotools/data/jdbc/datasource/JNDIDataSourceFactory.java:70]: Aside, if `getInitialContext()` declares it throws a `NamingException`, that's what should be catch. Any other thing will be a `RuntimeException` and hence totally unrelated. The message in the `DataSourceException` will be misleading, a runtime exception indicates some other kind of problem and needs a different treatment.

Comment (roarbra) [reply modules/library/jdbc/src/main/java/org/geotools/data/jdbc/datasource/JNDIDataSourceFactory.java:70]: I totally agree, but I'm not sure if I should temper this one at the moment. My only change is to remove the hints argument.

File: modules/library/metadata/src/main/java/org/geotools/util/factory/GeoTools.java (hunk 1 above)
Comment (roarbra) [reply modules/library/metadata/src/main/java/org/geotools/util/factory/GeoTools.java:1040]: That's ok.

**Pull request merged**